### ISSUE TITLE
[How PyTorch use mkldnn dispatch]

### ISSUE THREAD

**Issue by ddummkopfer** [oneapi-src/oneDNN#845]
Hi, I have a question about how **PyTorch** use **mkldnn** dispatch?
I use pip install torch, not build from torch source code.
I want to use the mkldnn dispatch op by the following code.
```
with torch.autograd.profiler.profile() as prof:
    input = torch.randn(10, 10, dtype=torch.float32).to_mkldnn()
    output = input.zero_()
print(prof.key_averages().table(sort_by="self_cpu_time_total"))  
```

The profiling result:
zero_                1.67%             
to_mkldnn      98.33%  

We all know the some op in PyTorch can be dispatched to some branch in file native_functions.yaml, like below.
```
- func: zero_(Tensor(a!) self) -> Tensor(a!)
  supports_named_tensor: True
  variants: method, function
  dispatch:
    CPU: zero_
    CUDA: zero_
    SparseCPU: zero_sparse_
    SparseCUDA: zero_sparse_
    MkldnnCPU: mkldnn_zero_
```

Why mkldnn_zero_ not be called? I set the DNNL_VERBOSE, but there is no log.
I wonder why it is.

**Comment by ddummkopfer** [comment]
Does that mean the profiling log only prints the func name in natives_functions.yaml, instead of the disptach name?
`- func: zero_(Tensor(a!) self) -> Tensor(a!) <--- Only print this layer
  supports_named_tensor: True
  variants: method, function
  dispatch:                                                   <---- Will not print this layer
    CPU: zero_ 
    CUDA: zero_
    SparseCPU: zero_sparse_
    SparseCUDA: zero_sparse_
    MkldnnCPU: mkldnn_zero_`

**Comment by emfomenk** [comment]
The question seems to be more PyTorch related than oneDNN.
Summon @pinzhenx and @CaoZhongZ to help.

**Comment by pinzhenx** [comment]
@ddummkopfer As `zero_` is not a computation operator provided by mkldnn, we are implementing `zero_` using aten vectorization facility directly [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mkldnn/MkldnnTensorMath.cpp#L34). Thus, you cannot see any log printed with `DNNL_VERBOSE`

**Comment by ddummkopfer** [comment]
Hi, @pinzhenx 

When I use batchnorm to do some calculation, why the BN is not dispatched to mkldnn version.

```
m = mkldnn_utils.to_mkldnn(nn.BatchNorm2d(100).eval())
input = torch.randn(20, 100, 35, 45).to_mkldnn()
output = m(input)

```
The profiling log:
_batch_norm_impl_index      0.16%            6.516us          45.00%           1.791ms          1.791ms          NaN              0.000us          
batch_norm                           0.20%            7.904us          45.20%           1.799ms          1.799ms          NaN              0.000us          
empty                                   0.25%            10.144us         0.25%            10.144us         10.144us         NaN              0.000us          
native_batch_norm               44.58%           1.775ms          44.58%           1.775ms          1.775ms          NaN              0.000us     
to_mkldnn                             54.30%           2.162ms          54.30%           2.162ms          432.359us        NaN              0.000us          

```
- func: native_batch_norm() -> (Tensor, Tensor, Tensor)
  dispatch:
    CPU: batch_norm_cpu
    CUDA: batch_norm_cuda
    MkldnnCPU: mkldnn_batch_norm
```
Does the profiling only print the func name, not print the dispatch func name.

**Comment by pinzhenx** [comment]
> Does the profiling only print the func name, not print the dispatch func name.

Yes. This time you can refer to the `DNNL_VERBOSE` to verify if it's dispatching into mkldnn correctly

**Comment by ddummkopfer** [comment]
Hi

I set DNNL_VERBOSE=1,=2, there is no log like yours.

```
>>> m = mkldnn_utils.to_mkldnn(nn.BatchNorm2d(100).eval())
>>>
```
```
>>> print(torch.__config__.show())
PyTorch built with:
  - GCC 7.5
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.1 Product Build 20200208 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_INTERNAL_THREADPOOL_IMPL -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=OFF, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,
```

**Comment by pinzhenx** [comment]
You are using an old version of pytorch built with mkdnn 0.21.

With mkldnn < 1.0, you shoud enable `MKLDNN_VERBOSE=1` to print log messages

**Comment by ddummkopfer** [comment]
@pinzhenx Thank you for your **significant** **help**. Now I can see the verbose log with `MKLDNN_VERBOSE=1`.
Another Question, I see there are native cpu operator in native_functions.yaml, how can I choose the native version instead of using mkldnn version.  I find the PyTorch will automatic choose mkldnn operator for my model, while i want to use native version. How do i need to do?

**Comment by pinzhenx** [comment]
Are you mentioning that pytorch dispatches convolution to the mkldnn?

In general, if you are not converting the input to mkldnn, namely calling `input.to_mkldnn()`, no mkldnn ops will be called. The only exception is Convolution, pytorch will choose mkldnn over their own thnn impl. in some conditions. You can easily disable MKLDNN conv at runtime as per https://github.com/oneapi-src/oneDNN/issues/791#issuecomment-665388955

**Comment by ddummkopfer** [comment]
@pinzhenx Great Thanks! Wonderful answer, Gain a lot from your answer. 👍 🥇 💯
